File: daedalos-tools/loop/lib/checkpoint.py
"""
Checkpoint backends for loop state preservation.

Every iteration creates a rollback point. This module provides:
- BtrfsCheckpoint: Instant, space-efficient snapshots (recommended)
- GitCheckpoint: Git-based checkpoints for non-Btrfs systems
- NoneCheckpoint: No checkpoints (for low-risk operations)
"""

from abc import ABC, abstractmethod
from pathlib import Path
import subprocess
import json
import shutil
from datetime import datetime
from typing import Optional, List
from dataclasses import dataclass


@dataclass
class Checkpoint:
    """Represents a checkpoint in the system."""
    id: str
    name: str
    created_at: str
    path: str
    backend: str
    metadata: dict


class CheckpointBackend(ABC):
    """Abstract base class for checkpoint backends."""

    @property
    @abstractmethod
    def name(self) -> str:
        """Return the backend name."""
        pass

    @abstractmethod
    def create(self, name: str, path: Path) -> str:
        """Create a checkpoint, return checkpoint ID."""
        pass

    @abstractmethod
    def restore(self, checkpoint_id: str, path: Path) -> bool:
        """Restore to a checkpoint."""
        pass

    @abstractmethod
    def list(self, path: Path) -> List[Checkpoint]:
        """List available checkpoints."""
        pass

    @abstractmethod
    def delete(self, checkpoint_id: str) -> bool:
        """Delete a checkpoint."""
        pass

    @abstractmethod
    def exists(self, checkpoint_id: str) -> bool:
        """Check if a checkpoint exists."""
        pass


class BtrfsCheckpoint(CheckpointBackend):
    """
    Btrfs snapshot-based checkpoints.

    Advantages:
    - Instant snapshots (< 1ms)
    - Copy-on-write (minimal space usage)
    - Full filesystem state captured
    - Works with any file type
    """

    def __init__(self, snapshot_dir: Optional[Path] = None):
        self.snapshot_dir = snapshot_dir or (
            Path.home() / ".local/share/daedalos/loop/snapshots"
        )
        self.snapshot_dir.mkdir(parents=True, exist_ok=True)
        self._metadata_file = self.snapshot_dir / "metadata.json"
        self._metadata = self._load_metadata()

    def _load_metadata(self) -> dict:
        """Load checkpoint metadata from JSON file."""
        if self._metadata_file.exists():
            try:
                with open(self._metadata_file) as f:
                    return json.load(f)
            except (json.JSONDecodeError, IOError):
                return {}
        return {}

    def _save_metadata(self):
        """Save checkpoint metadata to JSON file."""
        with open(self._metadata_file, "w") as f:
            json.dump(self._metadata, f, indent=2)

    @property
    def name(self) -> str:
        return "btrfs"

    def create(self, name: str, path: Path) -> str:
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        checkpoint_id = f"{name}_{timestamp}"
        snapshot_path = self.snapshot_dir / checkpoint_id

        try:
            result = subprocess.run(
                ["btrfs", "subvolume", "snapshot", "-r", str(path), str(snapshot_path)],
                check=True,
                capture_output=True,
                text=True
            )

            # Store metadata
            self._metadata[checkpoint_id] = {
                "name": name,
                "created_at": datetime.now().isoformat(),
                "source_path": str(path),
                "snapshot_path": str(snapshot_path)
            }
            self._save_metadata()

            return checkpoint_id

        except subprocess.CalledProcessError as e:
            raise RuntimeError(f"Failed to create Btrfs snapshot: {e.stderr}")

    def restore(self, checkpoint_id: str, path: Path) -> bool:
        snapshot_path = self.snapshot_dir / checkpoint_id
        if not snapshot_path.exists():
            return False

        try:
            # For Btrfs restore, we need to:
            # 1. Delete current subvolume (if it is one)
            # 2. Create new snapshot from checkpoint

            # Check if path is a subvolume
            check_result = subprocess.run(
                ["btrfs", "subvolume", "show", str(path)],
                capture_output=True
            )

            if check_result.returncode == 0:
                # It's a subvolume, delete it
                subprocess.run(
                    ["btrfs", "subvolume", "delete", str(path)],
                    check=True,
                    capture_output=True
                )

            # Create writable snapshot from the read-only checkpoint
            subprocess.run(
                ["btrfs", "subvolume", "snapshot", str(snapshot_path), str(path)],
                check=True,
                capture_output=True
            )

            return True

        except subprocess.CalledProcessError:
            return False

    def list(self, path: Path) -> List[Checkpoint]:
        checkpoints = []
        for checkpoint_id, meta in self._metadata.items():
            snapshot_path = self.snapshot_dir / checkpoint_id
            if snapshot_path.exists():
                checkpoints.append(Checkpoint(
                    id=checkpoint_id,
                    name=meta.get("name", checkpoint_id),
                    created_at=meta.get("created_at", ""),
                    path=str(snapshot_path),
                    backend="btrfs",
                    metadata=meta
                ))
        return sorted(checkpoints, key=lambda c: c.created_at, reverse=True)

    def delete(self, checkpoint_id: str) -> bool:
        snapshot_path = self.snapshot_dir / checkpoint_id
        if snapshot_path.exists():
            try:
                subprocess.run(
                    ["btrfs", "subvolume", "delete", str(snapshot_path)],
                    check=True,
                    capture_output=True
                )
                if checkpoint_id in self._metadata:
                    del self._metadata[checkpoint_id]
                    self._save_metadata()
                return True
            except subprocess.CalledProcessError:
                return False
        return False

    def exists(self, checkpoint_id: str) -> bool:
        return (self.snapshot_dir / checkpoint_id).exists()


class GitCheckpoint(CheckpointBackend):
    """
    Git-based checkpoints for non-Btrfs systems.

    Advantages:
    - Works everywhere git works
    - Familiar model for developers
    - Easy to inspect and compare

    Limitations:
    - Only tracks git-tracked files
    - Slower for large repos
    """

    def __init__(self, repo_path: Optional[Path] = None):
        self.repo_path = repo_path
        self._branch_prefix = "loop-checkpoint"

    def _git(self, *args, cwd: Optional[Path] = None) -> subprocess.CompletedProcess:
        """Run a git command."""
        working_dir = cwd or self.repo_path
        return subprocess.run(
            ["git", "-C", str(working_dir)] + list(args),
            capture_output=True,
            text=True
        )

    def _ensure_repo(self, path: Path):
        """Ensure we're working with a git repo."""
        if self.repo_path is None:
            self.repo_path = path

    @property
    def name(self) -> str:
        return "git"

    def create(self, name: str, path: Path) -> str:
        self._ensure_repo(path)
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        checkpoint_id = f"{self._branch_prefix}/{name}_{timestamp}"

        # First, stash any uncommitted changes
        stash_result = self._git("stash", "push", "-m", f"loop-auto-stash-{checkpoint_id}")
        had_changes = "No local changes" not in stash_result.stdout

        # Get current HEAD
        head_result = self._git("rev-parse", "HEAD")
        if head_result.returncode != 0:
            # No commits yet, create initial commit
            self._git("add", "-A")
            self._git("commit", "-m", "Initial commit for loop checkpoint", "--allow-empty")

        # Create checkpoint branch at current HEAD
        result = self._git("branch", checkpoint_id)
        if result.returncode != 0:
            raise RuntimeError(f"Failed to create checkpoint branch: {result.stderr}")

        # Restore stashed changes if any
        if had_changes:
            self._git("stash", "pop")

        return checkpoint_id

    def restore(self, checkpoint_id: str, path: Path) -> bool:
        self._ensure_repo(path)

        # Stash current changes
        self._git("stash", "push", "-m", "loop-pre-restore")

        # Get current branch name to return to after restore
        current_branch = self._git("rev-parse", "--abbrev-ref", "HEAD").stdout.strip()

        # Hard reset to checkpoint
        # First checkout the checkpoint
        result = self._git("checkout", checkpoint_id)
        if result.returncode != 0:
            self._git("stash", "pop")
            return False

        # If we were on a different branch, delete it and recreate at this point
        if current_branch != checkpoint_id and not current_branch.startswith(self._branch_prefix):
            self._git("branch", "-D", current_branch)
            self._git("checkout", "-b", current_branch)

        return True

    def list(self, path: Path) -> List[Checkpoint]:
        self._ensure_repo(path)
        result = self._git(
            "branch", "--list", f"{self._branch_prefix}/*",
            "--format=%(refname:short)|%(creatordate:iso-strict)"
        )

        checkpoints = []
        for line in result.stdout.strip().split("\n"):
            if "|" in line:
                branch, date = line.split("|", 1)
                # Extract name from branch
                name = branch.replace(f"{self._branch_prefix}/", "")
                checkpoints.append(Checkpoint(
                    id=branch,
                    name=name,
                    created_at=date,
                    path=str(self.repo_path),
                    backend="git",
                    metadata={"branch": branch}
                ))

        return sorted(checkpoints, key=lambda c: c.created_at, reverse=True)

    def delete(self, checkpoint_id: str) -> bool:
        result = self._git("branch", "-D", checkpoint_id)
        return result.returncode == 0

    def exists(self, checkpoint_id: str) -> bool:
        result = self._git("rev-parse", "--verify", checkpoint_id)
        return result.returncode == 0


class NoneCheckpoint(CheckpointBackend):
    """
    No-op checkpoint backend.

    Use for:
    - Low-risk operations
    - When speed is critical
    - Testing without side effects
    """

    @property
    def name(self) -> str:
        return "none"

    def create(self, name: str, path: Path) -> str:
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        return f"none_{name}_{timestamp}"

    def restore(self, checkpoint_id: str, path: Path) -> bool:
        # No-op: can't restore without checkpoints
        return False

    def list(self, path: Path) -> List[Checkpoint]:
        return []

    def delete(self, checkpoint_id: str) -> bool:
        return True

    def exists(self, checkpoint_id: str) -> bool:
        return False


def detect_backend(path: Path) -> str:
    """
    Auto-detect the best checkpoint backend for a path.

    Priority:
    1. Btrfs (if available and path is on Btrfs)
    2. Git (if path is in a git repo)
    3. None (fallback)
    """
    # Check for Btrfs
    try:
        result = subprocess.run(
            ["btrfs", "subvolume", "show", str(path)],
            capture_output=True
        )
        if result.returncode == 0:
            return "btrfs"
    except FileNotFoundError:
        pass  # btrfs command not available

    # Check for Git
    try:
        result = subprocess.run(
            ["git", "-C", str(path), "rev-parse", "--git-dir"],
            capture_output=True
        )
        if result.returncode == 0:
            return "git"
    except FileNotFoundError:
        pass  # git command not available

    return "none"


def get_backend(path: Path, strategy: str = "auto") -> CheckpointBackend:
    """
    Factory function to get appropriate checkpoint backend.

    Args:
        path: Working directory for the loop
        strategy: "auto", "btrfs", "git", or "none"

    Returns:
        Appropriate CheckpointBackend instance
    """
    if strategy == "auto":
        strategy = detect_backend(path)

    if strategy == "btrfs":
        return BtrfsCheckpoint()
    elif strategy == "git":
        return GitCheckpoint(path)
    elif strategy == "none":
        return NoneCheckpoint()
    else:
        raise ValueError(f"Unknown checkpoint strategy: {strategy}")

File: crates/loop/src/checkpoint.rs
//! Checkpoint backends for loop state preservation
//!
//! Every iteration creates a rollback point. This module provides:
//! - GitCheckpoint: Git-based checkpoints (works everywhere)
//! - BtrfsCheckpoint: Instant, space-efficient snapshots (Linux + Btrfs)
//! - NoneCheckpoint: No checkpoints (for low-risk operations)

use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use daedalos_core::Paths;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// Checkpoint strategy
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CheckpointStrategy {
    Auto,
    Git,
    Btrfs,
    None,
}

/// Represents a checkpoint
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Checkpoint {
    pub id: String,
    pub name: String,
    pub created_at: DateTime<Utc>,
    pub path: String,
    pub backend: String,
}

/// Trait for checkpoint backends
pub trait CheckpointBackend: Send + Sync {
    /// Return the backend name
    fn name(&self) -> &str;

    /// Create a checkpoint, return checkpoint ID
    fn create(&self, name: &str, path: &PathBuf) -> Result<String>;

    /// Restore to a checkpoint
    fn restore(&self, checkpoint_id: &str, path: &PathBuf) -> Result<bool>;

    /// List available checkpoints
    fn list(&self, path: &PathBuf) -> Result<Vec<Checkpoint>>;

    /// Delete a checkpoint
    fn delete(&self, checkpoint_id: &str) -> Result<bool>;

    /// Check if a checkpoint exists
    fn exists(&self, checkpoint_id: &str) -> bool;
}

/// Git-based checkpoints
///
/// Works everywhere git works. Uses branches to store checkpoints.
pub struct GitCheckpoint {
    repo_path: Option<PathBuf>,
    branch_prefix: String,
}

impl GitCheckpoint {
    pub fn new(repo_path: Option<PathBuf>) -> Self {
        Self {
            repo_path,
            branch_prefix: "loop-checkpoint".to_string(),
        }
    }

    fn git(&self, args: &[&str], cwd: Option<&PathBuf>) -> Result<std::process::Output> {
        let working_dir = cwd.or(self.repo_path.as_ref());
        let mut cmd = std::process::Command::new("git");

        if let Some(dir) = working_dir {
            cmd.arg("-C").arg(dir);
        }

        for arg in args {
            cmd.arg(arg);
        }

        cmd.output().context("Failed to run git command")
    }
}

impl CheckpointBackend for GitCheckpoint {
    fn name(&self) -> &str {
        "git"
    }

    fn create(&self, name: &str, path: &PathBuf) -> Result<String> {
        let timestamp = Utc::now().format("%Y%m%d_%H%M%S");
        let checkpoint_id = format!("{}/{}_{}", self.branch_prefix, name, timestamp);

        // Stash any uncommitted changes
        let stash_result = self.git(&["stash", "push", "-m", &format!("loop-auto-stash-{}", checkpoint_id)], Some(path))?;
        let had_changes = !String::from_utf8_lossy(&stash_result.stdout).contains("No local changes");

        // Check if we have any commits
        let head_result = self.git(&["rev-parse", "HEAD"], Some(path))?;
        if !head_result.status.success() {
            // No commits yet, create initial commit
            self.git(&["add", "-A"], Some(path))?;
            self.git(&["commit", "-m", "Initial commit for loop checkpoint", "--allow-empty"], Some(path))?;
        }

        // Create checkpoint branch
        let result = self.git(&["branch", &checkpoint_id], Some(path))?;
        if !result.status.success() {
            let stderr = String::from_utf8_lossy(&result.stderr);
            anyhow::bail!("Failed to create checkpoint branch: {}", stderr);
        }

        // Restore stashed changes
        if had_changes {
            self.git(&["stash", "pop"], Some(path))?;
        }

        Ok(checkpoint_id)
    }

    fn restore(&self, checkpoint_id: &str, path: &PathBuf) -> Result<bool> {
        // Stash current changes
        self.git(&["stash", "push", "-m", "loop-pre-restore"], Some(path))?;

        // Get current branch
        let current_result = self.git(&["rev-parse", "--abbrev-ref", "HEAD"], Some(path))?;
        let current_branch = String::from_utf8_lossy(&current_result.stdout).trim().to_string();

        // Checkout the checkpoint
        let result = self.git(&["checkout", checkpoint_id], Some(path))?;
        if !result.status.success() {
            self.git(&["stash", "pop"], Some(path))?;
            return Ok(false);
        }

        // If we were on a different branch, recreate it at this point
        if current_branch != checkpoint_id && !current_branch.starts_with(&self.branch_prefix) {
            self.git(&["branch", "-D", &current_branch], Some(path))?;
            self.git(&["checkout", "-b", &current_branch], Some(path))?;
        }

        Ok(true)
    }

    fn list(&self, path: &PathBuf) -> Result<Vec<Checkpoint>> {
        let result = self.git(
            &[
                "branch",
                "--list",
                &format!("{}/*", self.branch_prefix),
                "--format=%(refname:short)|%(creatordate:iso-strict)",
            ],
            Some(path),
        )?;

        let mut checkpoints = Vec::new();
        let output = String::from_utf8_lossy(&result.stdout);

        for line in output.lines() {
            if let Some((branch, date)) = line.split_once('|') {
                let name = branch
                    .strip_prefix(&format!("{}/", self.branch_prefix))
                    .unwrap_or(branch)
                    .to_string();

                let created_at = DateTime::parse_from_rfc3339(date)
                    .map(|d| d.with_timezone(&Utc))
                    .unwrap_or_else(|_| Utc::now());

                checkpoints.push(Checkpoint {
                    id: branch.to_string(),
                    name,
                    created_at,
                    path: path.to_string_lossy().to_string(),
                    backend: "git".to_string(),
                });
            }
        }

        checkpoints.sort_by(|a, b| b.created_at.cmp(&a.created_at));
        Ok(checkpoints)
    }

    fn delete(&self, checkpoint_id: &str) -> Result<bool> {
        let result = self.git(&["branch", "-D", checkpoint_id], None)?;
        Ok(result.status.success())
    }

    fn exists(&self, checkpoint_id: &str) -> bool {
        self.git(&["rev-parse", "--verify", checkpoint_id], None)
            .map(|r| r.status.success())
            .unwrap_or(false)
    }
}

/// Btrfs snapshot-based checkpoints
///
/// Instant snapshots with copy-on-write (minimal space usage).
/// Only works on Linux with Btrfs filesystem.
pub struct BtrfsCheckpoint {
    snapshot_dir: PathBuf,
}

impl BtrfsCheckpoint {
    pub fn new() -> Self {
        let paths = Paths::new();
        let snapshot_dir = paths.state("loop").join("snapshots");
        std::fs::create_dir_all(&snapshot_dir).ok();

        Self { snapshot_dir }
    }
}

impl CheckpointBackend for BtrfsCheckpoint {
    fn name(&self) -> &str {
        "btrfs"
    }

    fn create(&self, name: &str, path: &PathBuf) -> Result<String> {
        let timestamp = Utc::now().format("%Y%m%d_%H%M%S");
        let checkpoint_id = format!("{}_{}", name, timestamp);
        let snapshot_path = self.snapshot_dir.join(&checkpoint_id);

        let output = std::process::Command::new("btrfs")
            .args(["subvolume", "snapshot", "-r"])
            .arg(path)
            .arg(&snapshot_path)
            .output()
            .context("Failed to run btrfs command")?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("Failed to create Btrfs snapshot: {}", stderr);
        }

        Ok(checkpoint_id)
    }

    fn restore(&self, checkpoint_id: &str, path: &PathBuf) -> Result<bool> {
        let snapshot_path = self.snapshot_dir.join(checkpoint_id);
        if !snapshot_path.exists() {
            return Ok(false);
        }

        // Check if path is a subvolume
        let check_result = std::process::Command::new("btrfs")
            .args(["subvolume", "show"])
            .arg(path)
            .output()?;

        if check_result.status.success() {
            // Delete the subvolume
            std::process::Command::new("btrfs")
                .args(["subvolume", "delete"])
                .arg(path)
                .output()?;
        }

        // Create writable snapshot from the read-only checkpoint
        let result = std::process::Command::new("btrfs")
            .args(["subvolume", "snapshot"])
            .arg(&snapshot_path)
            .arg(path)
            .output()?;

        Ok(result.status.success())
    }

    fn list(&self, _path: &PathBuf) -> Result<Vec<Checkpoint>> {
        let mut checkpoints = Vec::new();

        if self.snapshot_dir.exists() {
            for entry in std::fs::read_dir(&self.snapshot_dir)? {
                let entry = entry?;
                let path = entry.path();
                if path.is_dir() {
                    let name = path.file_name().unwrap().to_string_lossy().to_string();
                    let metadata = std::fs::metadata(&path)?;
                    let created_at = metadata
                        .created()
                        .map(|t| DateTime::<Utc>::from(t))
                        .unwrap_or_else(|_| Utc::now());

                    checkpoints.push(Checkpoint {
                        id: name.clone(),
                        name,
                        created_at,
                        path: path.to_string_lossy().to_string(),
                        backend: "btrfs".to_string(),
                    });
                }
            }
        }

        checkpoints.sort_by(|a, b| b.created_at.cmp(&a.created_at));
        Ok(checkpoints)
    }

    fn delete(&self, checkpoint_id: &str) -> Result<bool> {
        let snapshot_path = self.snapshot_dir.join(checkpoint_id);
        if snapshot_path.exists() {
            let result = std::process::Command::new("btrfs")
                .args(["subvolume", "delete"])
                .arg(&snapshot_path)
                .output()?;
            Ok(result.status.success())
        } else {
            Ok(false)
        }
    }

    fn exists(&self, checkpoint_id: &str) -> bool {
        self.snapshot_dir.join(checkpoint_id).exists()
    }
}

/// No-op checkpoint backend
///
/// Use for low-risk operations or when speed is critical.
pub struct NoneCheckpoint;

impl NoneCheckpoint {
    pub fn new() -> Self {
        Self
    }
}

impl CheckpointBackend for NoneCheckpoint {
    fn name(&self) -> &str {
        "none"
    }

    fn create(&self, name: &str, _path: &PathBuf) -> Result<String> {
        let timestamp = Utc::now().format("%Y%m%d_%H%M%S");
        Ok(format!("none_{}_{}", name, timestamp))
    }

    fn restore(&self, _checkpoint_id: &str, _path: &PathBuf) -> Result<bool> {
        // Can't restore without checkpoints
        Ok(false)
    }

    fn list(&self, _path: &PathBuf) -> Result<Vec<Checkpoint>> {
        Ok(Vec::new())
    }

    fn delete(&self, _checkpoint_id: &str) -> Result<bool> {
        Ok(true)
    }

    fn exists(&self, _checkpoint_id: &str) -> bool {
        false
    }
}

/// Auto-detect the best checkpoint backend for a path
pub fn detect_backend(path: &PathBuf) -> CheckpointStrategy {
    // Check for Btrfs
    let btrfs_result = std::process::Command::new("btrfs")
        .args(["subvolume", "show"])
        .arg(path)
        .output();

    if let Ok(result) = btrfs_result {
        if result.status.success() {
            return CheckpointStrategy::Btrfs;
        }
    }

    // Check for Git
    let git_result = std::process::Command::new("git")
        .args(["-C"])
        .arg(path)
        .args(["rev-parse", "--git-dir"])
        .output();

    if let Ok(result) = git_result {
        if result.status.success() {
            return CheckpointStrategy::Git;
        }
    }

    CheckpointStrategy::None
}

/// Factory function to get appropriate checkpoint backend
pub fn get_backend(
    path: &PathBuf,
    strategy: CheckpointStrategy,
) -> Result<Box<dyn CheckpointBackend>> {
    let strategy = if strategy == CheckpointStrategy::Auto {
        detect_backend(path)
    } else {
        strategy
    };

    match strategy {
        CheckpointStrategy::Btrfs => Ok(Box::new(BtrfsCheckpoint::new())),
        CheckpointStrategy::Git => Ok(Box::new(GitCheckpoint::new(Some(path.clone())))),
        CheckpointStrategy::None | CheckpointStrategy::Auto => Ok(Box::new(NoneCheckpoint::new())),
    }
}
